File: 무인도여행.py
from collections import deque

def solution(maps):
    R, C = len(maps), len(maps[0])
    visited = [[False for j in range(C)] for i in range(R)]

    
    def bfs(y, x):
        # 상하좌우 이동하며 방문
        
        dy = [-1, 0, 1, 0]
        dx = [0, 1, 0, -1]
        
        q = deque([(y, x)])
        visited[y][x] = True
        day = 0
        
        while q:
            r, c = q.popleft()
            day += int(maps[r][c])
            
            
            for i in range(4):
                movey, movex = dy[i] + r, dx[i] + c
                if movey >= R or movex >= C or movey < 0 or movex < 0:
                    continue
                
                
                if maps[movey][movex] != 'X' and not visited[movey][movex]:
                    q.append((movey, movex))
                    visited[movey][movex] = True
        
        return day
            
    # BFS
    # 방문 가능한 노드 찾기
    
    answer = []
    for i in range(R):
        for j in range(C):
            if maps[i][j] != 'X' and not visited[i][j]:
                answer.append(bfs(i, j))
    
    
    return sorted(answer) if len(answer) else [-1]

File: 무인도여행.java
import java.util.*;


class Solution {
    private int r;
    private int c;
    
    private int[] dy = {-1, 0, 1, 0};
    private int[] dx = {0, 1, 0, -1};
    
    private int dfs(int y, int x, char[][] maze){
        int ans = 0;
        
        
        for(int i=0; i<4; i++){
            int movey = y + dy[i], movex = x + dx[i];
            
            if(movey < 0 || movey >= r || movex < 0 || movex >= c)
                continue;
            
            if(maze[movey][movex] != 'X'){
                int tmp = maze[movey][movex] - '0';
                maze[movey][movex] = 'X';
                
                ans += dfs(movey, movex, maze) + tmp;
            }
        }
        
        
        return ans;
        
        
    }
    
    public int[] solution(String[] maps) {
        r = maps.length; c = maps[0].length();
        
        // init
        char[][] maze = new char[r][c];
        for(int i=0; i<r; i++)
            maze[i] = maps[i].toCharArray();
        
        
        // dfs
        List<Integer> answer = new ArrayList<>();
        for(int i=0; i<r; i++){   
            for(int j=0; j<c; j++){
                if(maze[i][j] != 'X'){
                    int tmp = maze[i][j] - '0';
                    maze[i][j] = 'X';
                    
                    answer.add(dfs(i, j, maze) + tmp);
                }
            }
        }
        
        
        int[] ans = answer.stream().mapToInt(Integer:: intValue).toArray();
        
        Arrays.sort(ans);
        
        return ans.length == 0 ? new int[]{-1} : ans;
    }
}
